### PR TITLE
Add support for elastic jobs

### PR DESCRIPTION
**Pull request by milroy**
This PR lays the foundation for elastic job support.  It consists of three contributions:

1. A small amount of additional metadata to identify a job's type ("rigid" by default, optionally "elastic").
2. Changes to traverser to allow for the bypassing of elastic job insertion into planner.  It also handles rigid and elastic job removal.
3. Nine tests for correct behavior of elastic jobs, including to make sure rigid jobs take precedence, and errors on unrecognized job types.

The changes here lead to deprioritization of resource vertices with allocated elastic jobs in the upcoming elastic matcher PR.  The complexities of this behavior will be discussed in the corresponding PR.  

This PR addresses issues #551, #576, #590, and #591.

### CODE REVIEW

**Comment by milroy** [comment]
The Ubuntu: clang-6.0 chain-lint Travis test is failing with a weird error:
```
t4011-elastic-jobs.t: PASS: N=3   PASS=3   FAIL=0 SKIP=0 XPASS=0 XFAIL=0
ERROR: t4011-elastic-jobs.t - missing test plan
ERROR: t4011-elastic-jobs.t - exited with status 1
```
I'll investigate, but this appears to be a testsuite configuration error.

**Comment by grondo** [comment]
Try running `./t4011-elastic-jobs -d -v --chain-lint`

From the test-suite.log:
```
error: bug in the test script: broken &&-chain: 
    cat >cmds004 <<-EOF &&
	match allocate ${unit_job}
	match allocate ${unit_job}
	match allocate ${unit_job}
	match allocate ${unit_job}
	quit
EOF
```

**Comment by milroy** [comment]
> Try running `./t4011-elastic-jobs -d -v --chain-lint` [...]

Thanks @grondo!  I missed a `&&` as the message indicates.

File: resource/traversers/dfu_impl_update.cpp [hunk 1]
```diff
@@ -156,9 +156,16 @@ int dfu_impl_t::upd_plan (vtx_t u, const subsystem_t &s, unsigned int needs,
...
         if (jobmeta.allocate) {
             (*m_graph)[u].schedule.allocations[jobmeta.jobid] = span;
-            // update the allocated elastic job bool
-            (*m_graph)[u].schedule.elastic_job = (jobmeta.job_type == 
-                                                 "elastic") ? true : false;
+            // update the allocated elastic job metadata
+            if (jobmeta.job_type == "elastic") {
+                (*m_graph)[u].schedule.elastic_job = true;
+                (*m_graph)[u].schedule.elastic_at = jobmeta.at;
+                (*m_graph)[u].schedule.elastic_at = jobmeta.elastic_duration;
```
Comment (SteVwonder) [inline resource/traversers/dfu_impl_update.cpp]:
```suggestion
                (*m_graph)[u].schedule.elastic_duration = jobmeta.elastic_duration;
```

File: resource/traversers/dfu_impl_update.cpp [hunk 2]
```diff
@@ -156,9 +156,16 @@ int dfu_impl_t::upd_plan (vtx_t u, const subsystem_t &s, unsigned int needs,
...
+            } else {
+                (*m_graph)[u].schedule.elastic_job = false;
+                (*m_graph)[u].schedule.elastic_at = -1;
+                (*m_graph)[u].schedule.elastic_at = 0;
```
Comment (SteVwonder) [inline resource/traversers/dfu_impl_update.cpp]:
```suggestion
                (*m_graph)[u].schedule.elastic_duration = 0;                
```

**Comment by milroy** [comment]
Thanks @dongahn and @SteVwonder for your initial offline feedback.  To note here: we've determined that the Codecov relative scope coverage result is due to error checks changing line numbers, which is not indicative of a problem.

Once Travis is green this PR is ready for your review.

**Comment by SteVwonder** [comment]
Thanks @milroy!  It looks like the github action to validate commits is failing because there is not a newline between the commit subject and the commit body.  Other than that, those commit messages look valid.

File: resource/traversers/dfu_impl.cpp [hunk 3]
```diff
@@ -558,6 +558,16 @@ int dfu_impl_t::dom_dfv (const jobmeta_t &meta, vtx_t u,
         m_err_msg += strerror (errno);
         m_err_msg += ".\n";
         goto done;
+    } else if ((meta.job_type == "elastic") && (*m_graph)[u].schedule.elastic_job) {
```
Comment (SteVwonder) [inline resource/traversers/dfu_impl.cpp]: Super minor nit: not sure about the proper style here but elsewhere it seems like there is a space between the paren.s

```suggestion
    } else if ( (meta.job_type == "elastic") && (*m_graph)[u].schedule.elastic_job) {
```

**Comment by milroy** [comment]
> Thanks @milroy! It looks like the github action to validate commits is failing because there is not a newline between the commit subject and the commit body. Other than that, those commit messages look valid.

Yes, I'd forgotten to leave a blank line between the commit subject and body.  The commit validator is happy now.

File: resource/schema/sched_data.hpp [hunk 4]
```diff
@@ -40,6 +40,10 @@ struct schedule_t {
     std::map<int64_t, int64_t> allocations;
     std::map<int64_t, int64_t> reservations;
     planner_t *plans = NULL;
+    bool elastic_job = false;
+    bool elastic_job_running_at = false;
```
[the remark this answers is not in the document]
Comment (milroy) [reply resource/schema/sched_data.hpp]: Yes, I'm checking the `elastic_job_running_at` in the upcoming elastic matcher PR.  The other way to accomplish the same thing is to add a parameter to `dom_finish_vtx`, but that requires changing the interface.

File: resource/traversers/dfu_impl.cpp (hunk 3 above)
Comment (milroy) [reply resource/traversers/dfu_impl.cpp]: Good catch, I'll make that change.

**Comment by milroy** [comment]
> Generally LGTM! Two comments below. I'm also going to be honest and say I got a little lost on the tests that used the `update allocate` `resource-query` command, mainly due to my own unfamiliarity with that functionality. So maybe we can chat briefly offline about that over coffee tomorrow?

I'm happy to walk you though the `update_allocate` changes tomorrow afternoon.

File: resource/schema/sched_data.hpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (SteVwonder) [reply resource/schema/sched_data.hpp]: Ok.  Makes sense.  Do you mind adding a comment above these bools to highlight the difference between them?

[the remark this answers is not in the document]
Comment (milroy) [reply resource/schema/sched_data.hpp]: That's a good idea- I'll be sure to add a comment before this gets merged.

**Comment by dongahn** [comment]
@milroy: good work! A couple of suggestions to make our reviews a bit easier. 

- The third commit is pretty big and probably can be split. Please consider splitting it to
    1. Test outputs (i.e., expected R files)
    2. Test input job spec files
    3. The test itself

- Try to make your commit message imperative. For example, instead of
"This commit enables elastic jobs by adding metadata to keep track of exclusive job types", 
"Enable enable elastic jobs by adding metadata to keep track of exclusive job types. This can make our logs appear more consistent.

File: resource/schema/sched_data.cpp [hunk 5]
```diff
@@ -50,7 +50,6 @@ schedule_t &schedule_t::operator= (const schedule_t &o)
 {
     int64_t base_time = 0;
     uint64_t duration = 0;
-    size_t len = 0;
```
Comment (dongahn) [inline resource/schema/sched_data.cpp]: Thank you for detecting this deadwood. This can go to its own commit.

File: resource/schema/sched_data.hpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (dongahn) [reply resource/schema/sched_data.hpp]: Can you use `elastic_job` instead of `elastic_job_running_at` within your upcoming match policy PR, though. My guess is you are using  `elastic_job_running_at` to determine whether `elastic_at` and `elastic_duration` are valid. But I can't see a reason why `elastic_job` cannot be used for this check...

File: resource/traversers/dfu_impl.hpp [hunk 6]
```diff
@@ -57,6 +57,7 @@ struct jobmeta_t {
     int64_t jobid = -1;
     int64_t at = -1;
     uint64_t duration = SYSTEM_DEFAULT_DURATION; // will need config ultimately
+    std::string job_type = "rigid";
```
Comment (dongahn) [inline resource/traversers/dfu_impl.hpp]: This will soon be updated with configuration file support for resource. I have a slight preference to add this "default" resource type into "resource/config/system_defaults.hpp".

File: resource/traversers/dfu_impl.hpp [hunk 7]
```diff
@@ -68,6 +69,11 @@ struct jobmeta_t {
             duration = SYSTEM_MAX_DURATION; // need config support ultimately
         else
             duration = (int64_t)jobspec.attributes.system.duration;
+
+        std::unordered_map<std::string, YAML::Node>::const_iterator jtype =
```
Comment (dongahn) [inline resource/traversers/dfu_impl.hpp]: I forgot. Has our canonical job spec RFC updated for this?

**Comment by dongahn** [comment]
It seems the dfu code changes in c0e600136741a2a0b36a3ef405ae141b468e89a1 would better fit the second commit (2d129a74e069e6098a8197c442370243998f554e)?

File: resource/traversers/dfu.cpp [hunk 8]
```diff
@@ -78,22 +78,27 @@ int dfu_traverser_t::schedule (Jobspec::Jobspec &jobspec,
...
-        t = meta.at + 1;
-        p = (*get_graph ())[root].idata.subplans.at (dom);
-        len = planner_multi_resources_len (p);
-        duration = meta.duration;
-        detail::dfu_impl_t::count_relevant_types (p, dfv, agg);
-        for (t = planner_multi_avail_time_first (p, t, duration, &agg[0], len);
-             (t != -1 && rc && !errno); t = planner_multi_avail_time_next (p)) {
-            meta.at = t;
-            rc = detail::dfu_impl_t::select (jobspec, root, meta, x);
+        if (meta.job_type == "rigid") {
```
Comment (dongahn) [inline resource/traversers/dfu.cpp]: Now that I think about this, this will have impact when we choose backfill scheduling policies like EASY or CONSERVATIVE. What's going to happen if a user submits a bunch of elastic jobs, and our backfill scheduler tried to reserve them to calculate backfill?

**Comment by dongahn** [comment]
It is too late and I am not thinking clearly. I will look at the rest of the PR tomorrow.

File: resource/traversers/dfu.cpp (hunk 8 above)
Comment (milroy) [reply resource/traversers/dfu.cpp]: My intention with `meta.job_type == "rigid"` is to protect against reserving elastic jobs by triggering an `EINVAL` when the check fails.  Will this not work as intended?

File: resource/traversers/dfu_impl.hpp (hunk 6 above)
Comment (milroy) [reply resource/traversers/dfu_impl.hpp]: I can certainly do that instead.

File: resource/traversers/dfu_impl.hpp (hunk 7 above)
Comment (milroy) [reply resource/traversers/dfu_impl.hpp]: No, but since I've placed `job_type` in `system` attributes the RFC doesn't need to change.  We could make a new attribute or modify another existing one if there's a consensus preference.

**Comment by milroy** [comment]
> The third commit is pretty big and probably can be split. Please consider splitting it to [...]

I'll split up the third commit as you indicate.

>Try to make your commit message imperative. For example, instead of
"This commit enables elastic jobs by adding metadata to keep track of exclusive job types",
"Enable enable elastic jobs by adding metadata to keep track of exclusive job types. This can make our logs appear more consistent.

That's valuable feedback, I'll change my messages and make sure to do that in the future.

File: resource/schema/sched_data.hpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (milroy) [reply resource/schema/sched_data.hpp]: I created `elastic_job_running_at` with the intention of simplifying the code additions to `dfu_impl_t::dom_dfv`.  Upon further thought, I can add a temp variable in `dfu_impl_t::dom_dfv` with a couple more lines of code to update and reset `elastic_job`.  That should work fine and I won't need `elastic_job_running_at`.

**Comment by milroy** [comment]
> It seems the dfu code changes in c0e6001 would better fit the second commit (2d129a7)?

@dongahn: yes, that makes sense.  I can put it into 2d129a7.

File: resource/traversers/dfu.cpp (hunk 8 above)
Comment (milroy) [reply resource/traversers/dfu.cpp]: I do have a test to make sure elastic jobs don't get reserved, but does that not take into account different backfill policies?

Comment (dongahn) [reply resource/traversers/dfu.cpp]: Yes I saw that. But I think the problem will occur when you hook this into `resource`. Consider the following scenario.

1. EASY is selected. 
2. an elastic job is the highest priority job which cannot be "allocated" so EASY will try to reserve it to the earliest time point (through `allocate_orelse_reserve`)
3. but this will get `EINVAL`, which means an error was encountered.

At this point, I think EASY won't know what to do.

Comment (milroy) [reply resource/traversers/dfu.cpp]: Ok, I see.  I hadn't considered how the backfill policies handle errors.

Comment (dongahn) [reply resource/traversers/dfu.cpp]: I think we need to go back to the drawing board and think about the relationship between elastic job support and queuing policies. 

One can imagine we limit our elastic job support to FCFS thereby only `allocate` is used and thus elastic job support is clearly defined. However, the concepts of queuing policies and jobspecs are completely orthogonal at this point. IOW, all policies are clearly defined for *any* jobspec and thus good separation of concerns. So I hate to make a special case like this. 

It would be best if the concept of elastic jobs is such that it is clearly defined against all of the queueing policies. Like it is with regular jobs. -- Seems like a worthwhile topic to discuss.

Comment (SteVwonder) [reply resource/traversers/dfu.cpp]: One thought, maybe we could combine this functionality with something like #588 so that `allocate` is called on a preemptible job and never `allocate_orelse_reserve`.  But you could still use EASY backfilling etc.

Another thought, we could leverage the multiple queue idea in https://github.com/flux-framework/flux-sched/issues/529 to allow multiple scheduling policies depending on the jobs label (e.g. `rigid`, `elastic`, `batch`, `debug`).

**Comment by milroy** [comment]
I'm closing this PR.  Some of the features might make it into future PRs so I'm keeping the branch on my fork.
